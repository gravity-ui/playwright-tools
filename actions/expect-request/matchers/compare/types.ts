import type { Json, JsonPrimitive } from '../../types';

export const DiffType = {
    ValueMismatch: 'value-mismatch',
    ObjectMissingProperty: 'object-missing-property',
    ArrayMissingValue: 'array-missing-value',
} as const;
export type DiffType = (typeof DiffType)[keyof typeof DiffType];

export type DiffLog = DiffItem[];

export type DiffItem = DiffItemBase & { context: DiffContextItem[] };
export type DiffItemBase = {
    side?: 'left' | 'right';
} & (
    | {
          type: typeof DiffType.ValueMismatch;
          expected?: Json;
          received?: Json;
      }
    | {
          type: typeof DiffType.ObjectMissingProperty;
          expected: { key: string; value: Json };
      }
    | {
          type: typeof DiffType.ArrayMissingValue;
          expected: Json;
      }
);

export type DiffContextItem = {
    objectField?: string;
    arrayIndex?: number;
};

export const DecoratedLineType = {
    ObjectOpen: 'object-open',
    ObjectClose: 'object-close',
    ArrayOpen: 'array-open',
    ArrayClose: 'array-close',
    PrimitiveValue: 'primitive-value',
    EmptyObject: 'empty-object',
    EmptyArray: 'empty-array',
} as const;
export type DecoratedLineType = (typeof DecoratedLineType)[keyof typeof DecoratedLineType];

export type DecoratedLine = {
    fieldName?: string;
    level: number;
    diffMarker?: 'expected' | 'received';
} & (
    | {
          type: typeof DecoratedLineType.PrimitiveValue;
          value: JsonPrimitive;
      }
    | {
          type: Exclude<DecoratedLineType, typeof DecoratedLineType.PrimitiveValue>;
      }
);

export type PathAnnotation =
    | ObjectMissingPropertyAnnotation
    | ObjectExtraPropertyAnnotation
    | ArrayMissingValueAnnotation
    | ValueMismatchAnnotation;

export const PathAnnotationType = {
    ValueMismatch: 'value-mismatch',
    ObjectMissingProperty: 'object-missing-property',
    ObjectExtraProperty: 'object-extra-property',
    ArrayMissingValue: 'array-missing-value',
} as const;
export type PathAnnotationType = (typeof PathAnnotationType)[keyof typeof PathAnnotationType];

type ObjectMissingPropertyAnnotation = {
    type: typeof PathAnnotationType.ObjectMissingProperty;
    expected: { key: string; value: Json };
};
type ObjectExtraPropertyAnnotation = {
    type: typeof PathAnnotationType.ObjectExtraProperty;
};
type ArrayMissingValueAnnotation = {
    type: typeof PathAnnotationType.ArrayMissingValue;
    expected: Json;
};
type ValueMismatchAnnotation = {
    type: typeof PathAnnotationType.ValueMismatch;
    expected?: Json;
    received?: Json;
};

export function isValueMismatchAnnotation(
    annotation: PathAnnotation,
): annotation is ValueMismatchAnnotation {
    return annotation.type === PathAnnotationType.ValueMismatch;
}

export function isObjectMissingPropertyAnnotation(
    annotation: PathAnnotation,
): annotation is ObjectMissingPropertyAnnotation {
    return annotation.type === PathAnnotationType.ObjectMissingProperty;
}

export function isObjectExtraPropertyAnnotation(
    annotation: PathAnnotation,
): annotation is ArrayMissingValueAnnotation {
    return annotation.type === PathAnnotationType.ObjectExtraProperty;
}

export function isArrayMissingValueAnnotation(
    annotation: PathAnnotation,
): annotation is ArrayMissingValueAnnotation {
    return annotation.type === PathAnnotationType.ArrayMissingValue;
}
