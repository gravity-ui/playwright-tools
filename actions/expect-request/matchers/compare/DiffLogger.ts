import type { DiffContextItem, DiffItemBase, DiffLog } from './constants';

export class DiffLogger {
    private context: DiffContextItem[] = [];
    private _diffLog: DiffLog = [];

    get diffLog() {
        return this._diffLog;
    }

    addDiffItem(diffItem: DiffItemBase) {
        this.diffLog.push({ ...diffItem, context: [...this.context] });
    }

    addContext(context: DiffContextItem) {
        this.context.push(context);
    }

    removeContext() {
        this.context.pop();
    }
}
