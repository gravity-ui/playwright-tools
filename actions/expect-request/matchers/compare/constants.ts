export const DiffType = {
    ValueMismatch: 'value-mismatch',
    ObjectMissingProperty: 'object-missing-property',
    ArrayMissingValue: 'array-missing-value',
} as const;

export const DecoratedLineType = {
    ObjectOpen: 'object-open',
    ObjectClose: 'object-close',
    ArrayOpen: 'array-open',
    ArrayClose: 'array-close',
    PrimitiveValue: 'primitive-value',
    EmptyObject: 'empty-object',
    EmptyArray: 'empty-array',
} as const;

export const PathAnnotationType = {
    ValueMismatch: 'value-mismatch',
    ObjectMissingProperty: 'object-missing-property',
    ObjectExtraProperty: 'object-extra-property',
    ArrayMissingValue: 'array-missing-value',
} as const;
