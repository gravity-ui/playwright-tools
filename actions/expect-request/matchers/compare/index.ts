export { compare } from './compare';
export * from './types';
export { printJsonDiff } from './print-json-diff';
