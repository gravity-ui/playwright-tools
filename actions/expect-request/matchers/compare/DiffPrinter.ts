import type { Json } from '../../types';

import { compare } from './compare';
import { printJsonDiff } from './print-json-diff';
import type { DiffContextItem, DiffLog, PathAnnotation } from './constants';
import { DiffType, PathAnnotationType } from './constants';

interface DiffPrinterParams {
    left: Json;
    right: Json;
}

export class DiffPrinter {
    private left: Json;
    private right: Json;

    constructor({ left, right }: DiffPrinterParams) {
        this.left = left;
        this.right = right;
    }

    printDiff(params: { side: 'left' | 'right'; exact?: boolean }) {
        const { side, exact } = params;

        const diffLog = compare({ left: this.left, right: this.right, exactArrays: exact });
        const filteredLog = diffLog.reduce<DiffLog>((acc, logItem) => {
            if (logItem.side === side) {
                acc.push(logItem);
            } else if (exact && logItem.type === DiffType.ObjectMissingProperty) {
                acc.push({
                    type: DiffType.ValueMismatch,
                    received: logItem.expected.value,
                    side,
                    context: [...logItem.context, { objectField: logItem.expected.key }],
                });
            }

            return acc;
        }, []);

        if (!filteredLog.length) {
            return '';
        }

        const pathAnnotations = filteredLog.reduce<Record<string, PathAnnotation[]>>(
            (acc, logItem) => {
                const { context, type } = logItem;
                const path = this.makePathFromContext(context);

                if (!acc[path]) {
                    /* eslint-disable no-return-assign, no-param-reassign */
                    acc[path] = [];
                }

                if (type === DiffType.ValueMismatch) {
                    if (logItem.expected === undefined && logItem.received !== undefined) {
                        acc[path].push({ type: PathAnnotationType.ObjectExtraProperty });
                    } else {
                        acc[path].push({
                            type: PathAnnotationType.ValueMismatch,
                            expected: logItem.expected,
                            received: logItem.received,
                        });
                    }
                } else if (type === DiffType.ObjectMissingProperty) {
                    acc[path].push({
                        type: PathAnnotationType.ObjectMissingProperty,
                        expected: logItem.expected,
                    });
                } else if (type === DiffType.ArrayMissingValue) {
                    acc[path].push({
                        type: PathAnnotationType.ArrayMissingValue,
                        expected: logItem.expected,
                    });
                }

                return acc;
            },
            {},
        );

        return printJsonDiff({ json: side === 'left' ? this.left : this.right, pathAnnotations });
    }

    private makePathFromContext(context: DiffContextItem[]) {
        return context
            .map<string>(({ objectField, arrayIndex }) => {
                if (arrayIndex !== undefined) {
                    return `[${arrayIndex}]`;
                }

                return `${objectField}`;
            })
            .join('.');
    }
}
