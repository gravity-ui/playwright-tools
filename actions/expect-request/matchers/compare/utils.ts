import { PathAnnotationType } from './constants';
import type {
    ArrayMissingValueAnnotation,
    ObjectMissingPropertyAnnotation,
    PathAnnotation,
    ValueMismatchAnnotation,
} from './types';

export function isValueMismatchAnnotation(
    annotation: PathAnnotation,
): annotation is ValueMismatchAnnotation {
    return annotation.type === PathAnnotationType.ValueMismatch;
}

export function isObjectMissingPropertyAnnotation(
    annotation: PathAnnotation,
): annotation is ObjectMissingPropertyAnnotation {
    return annotation.type === PathAnnotationType.ObjectMissingProperty;
}

export function isObjectExtraPropertyAnnotation(
    annotation: PathAnnotation,
): annotation is ArrayMissingValueAnnotation {
    return annotation.type === PathAnnotationType.ObjectExtraProperty;
}

export function isArrayMissingValueAnnotation(
    annotation: PathAnnotation,
): annotation is ArrayMissingValueAnnotation {
    return annotation.type === PathAnnotationType.ArrayMissingValue;
}
