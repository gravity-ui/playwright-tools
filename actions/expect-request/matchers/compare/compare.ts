/* eslint-disable complexity */
import type { Json, JsonObject } from '../../types';
import { isObject } from '../utils';

import { DiffLogger } from './DiffLogger';
import { DiffType } from './constants';

export function compare(params: { left: Json; right: Json; exactArrays?: boolean }) {
    const { left, right, exactArrays } = params;

    const logger = new DiffLogger();

    compareInner({
        left,
        right,
        exactArrays,
        logger: logger,
    });

    compareInner({
        left: right,
        right: left,
        forwardComparison: false,
        exactArrays,
        logger: logger,
    });

    return logger.diffLog;
}

function compareInner(params: {
    left: Json;
    right: Json;
    forwardComparison?: boolean;
    exactArrays?: boolean;
    logger?: DiffLogger;
}): DiffLogger {
    const { left, right, exactArrays, logger, forwardComparison = true } = params;

    const innerLogger = logger ? logger : new DiffLogger();

    if (left === right) {
        return innerLogger;
    }

    if (left === null) {
        innerLogger.addDiffItem({
            type: DiffType.ValueMismatch,
            side: forwardComparison ? 'left' : 'right',
            expected: right,
            received: left,
        });

        return innerLogger;
    }

    if (Array.isArray(left) && Array.isArray(right)) {
        if (exactArrays) {
            if (left.length === right.length) {
                for (let i = 0; i < left.length; ++i) {
                    const leftItem = left[i] ?? null;
                    const rightItem = right[i] ?? null;

                    innerLogger.addContext({ arrayIndex: i });
                    compareInner({
                        left: leftItem,
                        right: rightItem,
                        forwardComparison: forwardComparison,
                        exactArrays,
                        logger: innerLogger,
                    });
                    innerLogger.removeContext();
                }
            } else {
                innerLogger.addDiffItem({
                    type: DiffType.ValueMismatch,
                    side: forwardComparison ? 'left' : 'right',
                    expected: right,
                    received: left,
                });
            }
        } else {
            const rightCopy = [...right];

            for (const leftItem of left) {
                const index = rightCopy.findIndex((rightItem) => {
                    const logger = compareInner({
                        left: leftItem,
                        right: rightItem,
                        exactArrays,
                        forwardComparison,
                    });

                    return logger.diffLog.length === 0;
                });

                if (index >= 0) {
                    rightCopy.splice(index, 1);
                }

                if (rightCopy.length === 0) {
                    break;
                }
            }

            for (const item of rightCopy) {
                innerLogger.addDiffItem({
                    type: DiffType.ArrayMissingValue,
                    side: forwardComparison ? 'left' : 'right',
                    expected: item,
                });
            }
        }
    } else if (isObject(left) && isObject(right)) {
        for (const [rightKey, rightValue] of Object.entries(right)) {
            innerLogger.addContext({ objectField: rightKey });

            if (!Object.hasOwn(left, rightKey)) {
                innerLogger.removeContext();

                innerLogger.addDiffItem({
                    type: DiffType.ObjectMissingProperty,
                    side: forwardComparison ? 'left' : 'right',
                    expected: {
                        key: rightKey,
                        value: rightValue,
                    },
                });

                continue;
            }

            const leftValue = (left as JsonObject)[rightKey] ?? null;

            compareInner({
                left: leftValue,
                right: rightValue,
                forwardComparison,
                exactArrays,
                logger: innerLogger,
            });

            innerLogger.removeContext();
        }
    } else if (left !== right) {
        innerLogger.addDiffItem({
            type: DiffType.ValueMismatch,
            side: forwardComparison ? 'left' : 'right',
            expected: right,
            received: left,
        });
    }

    return innerLogger;
}
