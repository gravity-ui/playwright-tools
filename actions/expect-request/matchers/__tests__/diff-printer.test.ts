import { describe, expect, it } from '@jest/globals';

import type { Json } from '../../types';
import { DiffPrinter } from '../compare/DiffPrinter';

function makeExpects({ left, right }: { left: Json; right: Json }) {
    const printer = new DiffPrinter({ left, right });

    const leftSideDiff = printer.printDiff({ side: 'left' });
    expect(leftSideDiff).toMatchSnapshot('1) left side diff');

    const rightSideDiff = printer.printDiff({ side: 'right' });
    expect(rightSideDiff).toMatchSnapshot('2) right side diff');

    const leftSideDiffExact = printer.printDiff({ side: 'left', exact: true });
    expect(leftSideDiffExact).toMatchSnapshot('3) left side diff exact');

    const rightSideDiffExact = printer.printDiff({ side: 'right', exact: true });
    expect(rightSideDiffExact).toMatchSnapshot('4) right side diff exact');
}

describe('diff-printer', () => {
    it('should not render diff when left and right are equal', () => {
        makeExpects({
            left: {
                arrayData: [1, 2, { textData: 'txt' }],
                textData: 'txt',
                nullData: null,
                booleanData: true,
                objectData: { textData: 'txt' },
            },
            right: {
                textData: 'txt',
                nullData: null,
                booleanData: true,
                objectData: { textData: 'txt' },
                arrayData: [1, 2, { textData: 'txt' }],
            },
        });
    });

    it('renders diff when left and right are primitives', () => {
        makeExpects({
            left: true,
            right: null,
        });
    });

    it('renders diff when left is primitive and right is empty object', () => {
        makeExpects({
            left: 'text',
            right: {},
        });
    });

    it('renders diff when left is primitive and right is empty array', () => {
        makeExpects({
            left: 'text',
            right: [],
        });
    });

    it('renders diff when left is empty object and right is empty array', () => {
        makeExpects({
            left: {},
            right: [],
        });
    });

    it('renders diff when left is primitive and right is object', () => {
        makeExpects({
            left: 'text',
            right: { data: 'text' },
        });
    });

    it('renders diff when left is object with empty object and right is primitive', () => {
        makeExpects({
            left: { emptyObject: {} },
            right: 'text',
        });
    });

    it('renders diff when left is object with empty array and right is primitive', () => {
        makeExpects({
            left: { emptyArray: [] },
            right: 'text',
        });
    });

    it('renders diff when left and right contain different empty non-primitives', () => {
        makeExpects({
            left: { emptyNonPrimitive: [] },
            right: { emptyNonPrimitive: {} },
        });
    });

    it('renders diff when right array includes left array', () => {
        makeExpects({
            left: [1, 2],
            right: [3, 1, 2],
        });
    });

    it('renders diff when left and right are arrays with equal length', () => {
        makeExpects({
            left: [1, 2, 3],
            right: [3, 2, 1],
        });
    });

    it('renders diff when left and right are objects', () => {
        makeExpects({
            left: { extraField: 'data', sameData: 'txt' },
            right: { rightField: 'data', sameData: 'txt' },
        });
    });

    it('renders diff when left is missing properties or right has extra properties', () => {
        makeExpects({
            left: { name: 'some-name', extraField: 'some-data' },
            right: {
                name: 'some-name',
                age: 33,
                objectData: { text: 'text' },
                nullData: null,
                arrayData: [1],
            },
        });
    });

    it('renders diff with deep nested primitive value mismatch', () => {
        makeExpects({
            left: { lvl1: { lvl2: { lvl3: { mismatch: 'data', sameData: 'txt' } } } },
            right: { lvl1: { lvl2: { lvl3: { mismatch: 'date', sameData: 'txt' } } } },
        });
    });

    it('renders diff with nested primitive and object mismatch', () => {
        makeExpects({
            left: { lvl1: { lvl2: { lvl3: { mismatch: 'data', sameData: 'txt' } } } },
            right: { lvl1: { lvl2: { lvl3: { mismatch: { data: 'txt' }, sameData: 'txt' } } } },
        });
    });

    it('renders diff with nested arrays with mismatches', () => {
        makeExpects({
            left: {
                data: [
                    [1, 2],
                    [3, 4],
                ],
            },
            right: {
                data: [
                    [1, 2],
                    [3, 5],
                ],
            },
        });
    });

    it('renders diff with special characters in keys', () => {
        makeExpects({
            left: { 'key-with-dash': 'value1', 'key.with.dots': 'value2' },
            right: { 'key-with-dash': 'different', 'key.with.dots': 'value2' },
        });
    });

    it('renders diff for complex nested structures', () => {
        makeExpects({
            left: {
                level1: {
                    level2A: {
                        level3A: { value: 'original' },
                        level3B: [1, 2, { nestedValue: 'same' }],
                    },
                    level2B: 'unchanged',
                },
            },
            right: {
                level1: {
                    level2A: {
                        level3A: { value: 'changed' },
                        level3B: [1, 2, { nestedValue: 'same' }],
                    },
                    level2B: 'unchanged',
                },
            },
        });
    });

    it('renders diff for arrays with mixed types', () => {
        makeExpects({
            left: [1, 'string', true, { key1: 'value1', key2: 'value2' }],
            right: [1, 'different', true, { key1: 'changed1', key2: 'changed2' }],
        });
    });

    it('renders diff for objects with numeric keys', () => {
        makeExpects({
            left: { '0': 'zero', '1': 'one' },
            right: { '0': 'zero', '1': 'changed' },
        });
    });
});
