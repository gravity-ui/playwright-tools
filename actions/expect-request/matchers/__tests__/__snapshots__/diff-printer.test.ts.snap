// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff-printer renders diff for arrays with mixed types: 1) left side diff 1`] = `
"  [
    1,
    "string",
    true,
    {
      "key1": "value1",
      "key2": "value2",
    },
-   "different",
-   {
-     "key1": "changed1",
-     "key2": "changed2",
-   },
  ]"
`;

exports[`diff-printer renders diff for arrays with mixed types: 2) right side diff 1`] = `
"  [
    1,
    "different",
    true,
    {
      "key1": "changed1",
      "key2": "changed2",
    },
-   "string",
-   {
-     "key1": "value1",
-     "key2": "value2",
-   },
  ]"
`;

exports[`diff-printer renders diff for arrays with mixed types: 3) left side diff exact 1`] = `
"  [
    1,
-   "different",
+   "string",
    true,
    {
-     "key1": "changed1",
+     "key1": "value1",
-     "key2": "changed2",
+     "key2": "value2",
    },
  ]"
`;

exports[`diff-printer renders diff for arrays with mixed types: 4) right side diff exact 1`] = `
"  [
    1,
-   "string",
+   "different",
    true,
    {
-     "key1": "value1",
+     "key1": "changed1",
-     "key2": "value2",
+     "key2": "changed2",
    },
  ]"
`;

exports[`diff-printer renders diff for complex nested structures: 1) left side diff 1`] = `
"  {
    "level1": {
      "level2A": {
        "level3A": {
-         "value": "changed",
+         "value": "original",
        },
        "level3B": [
          1,
          2,
          {
            "nestedValue": "same",
          },
        ],
      },
      "level2B": "unchanged",
    },
  }"
`;

exports[`diff-printer renders diff for complex nested structures: 2) right side diff 1`] = `
"  {
    "level1": {
      "level2A": {
        "level3A": {
-         "value": "original",
+         "value": "changed",
        },
        "level3B": [
          1,
          2,
          {
            "nestedValue": "same",
          },
        ],
      },
      "level2B": "unchanged",
    },
  }"
`;

exports[`diff-printer renders diff for complex nested structures: 3) left side diff exact 1`] = `
"  {
    "level1": {
      "level2A": {
        "level3A": {
-         "value": "changed",
+         "value": "original",
        },
        "level3B": [
          1,
          2,
          {
            "nestedValue": "same",
          },
        ],
      },
      "level2B": "unchanged",
    },
  }"
`;

exports[`diff-printer renders diff for complex nested structures: 4) right side diff exact 1`] = `
"  {
    "level1": {
      "level2A": {
        "level3A": {
-         "value": "original",
+         "value": "changed",
        },
        "level3B": [
          1,
          2,
          {
            "nestedValue": "same",
          },
        ],
      },
      "level2B": "unchanged",
    },
  }"
`;

exports[`diff-printer renders diff for objects with numeric keys: 1) left side diff 1`] = `
"  {
    "0": "zero",
-   "1": "changed",
+   "1": "one",
  }"
`;

exports[`diff-printer renders diff for objects with numeric keys: 2) right side diff 1`] = `
"  {
    "0": "zero",
-   "1": "one",
+   "1": "changed",
  }"
`;

exports[`diff-printer renders diff for objects with numeric keys: 3) left side diff exact 1`] = `
"  {
    "0": "zero",
-   "1": "changed",
+   "1": "one",
  }"
`;

exports[`diff-printer renders diff for objects with numeric keys: 4) right side diff exact 1`] = `
"  {
    "0": "zero",
-   "1": "one",
+   "1": "changed",
  }"
`;

exports[`diff-printer renders diff when left and right are arrays with equal length: 1) left side diff 1`] = `""`;

exports[`diff-printer renders diff when left and right are arrays with equal length: 2) right side diff 1`] = `""`;

exports[`diff-printer renders diff when left and right are arrays with equal length: 3) left side diff exact 1`] = `
"  [
-   3,
+   1,
    2,
-   1,
+   3,
  ]"
`;

exports[`diff-printer renders diff when left and right are arrays with equal length: 4) right side diff exact 1`] = `
"  [
-   1,
+   3,
    2,
-   3,
+   1,
  ]"
`;

exports[`diff-printer renders diff when left and right are objects: 1) left side diff 1`] = `
"  {
-   "rightField": "data",
    "extraField": "data",
    "sameData": "txt",
  }"
`;

exports[`diff-printer renders diff when left and right are objects: 2) right side diff 1`] = `
"  {
-   "extraField": "data",
    "rightField": "data",
    "sameData": "txt",
  }"
`;

exports[`diff-printer renders diff when left and right are objects: 3) left side diff exact 1`] = `
"  {
-   "rightField": "data",
+   "extraField": "data",
    "sameData": "txt",
  }"
`;

exports[`diff-printer renders diff when left and right are objects: 4) right side diff exact 1`] = `
"  {
-   "extraField": "data",
+   "rightField": "data",
    "sameData": "txt",
  }"
`;

exports[`diff-printer renders diff when left and right are primitives: 1) left side diff 1`] = `
"- null
+ true"
`;

exports[`diff-printer renders diff when left and right are primitives: 2) right side diff 1`] = `
"- true
+ null"
`;

exports[`diff-printer renders diff when left and right are primitives: 3) left side diff exact 1`] = `
"- null
+ true"
`;

exports[`diff-printer renders diff when left and right are primitives: 4) right side diff exact 1`] = `
"- true
+ null"
`;

exports[`diff-printer renders diff when left and right contain different empty non-primitives: 1) left side diff 1`] = `
"  {
-   "emptyNonPrimitive": {},
+   "emptyNonPrimitive": [],
  }"
`;

exports[`diff-printer renders diff when left and right contain different empty non-primitives: 2) right side diff 1`] = `
"  {
-   "emptyNonPrimitive": [],
+   "emptyNonPrimitive": {},
  }"
`;

exports[`diff-printer renders diff when left and right contain different empty non-primitives: 3) left side diff exact 1`] = `
"  {
-   "emptyNonPrimitive": {},
+   "emptyNonPrimitive": [],
  }"
`;

exports[`diff-printer renders diff when left and right contain different empty non-primitives: 4) right side diff exact 1`] = `
"  {
-   "emptyNonPrimitive": [],
+   "emptyNonPrimitive": {},
  }"
`;

exports[`diff-printer renders diff when left is empty object and right is empty array: 1) left side diff 1`] = `
"- []
+ {}"
`;

exports[`diff-printer renders diff when left is empty object and right is empty array: 2) right side diff 1`] = `"  []"`;

exports[`diff-printer renders diff when left is empty object and right is empty array: 3) left side diff exact 1`] = `
"- []
+ {}"
`;

exports[`diff-printer renders diff when left is empty object and right is empty array: 4) right side diff exact 1`] = `"  []"`;

exports[`diff-printer renders diff when left is missing properties or right has extra properties: 1) left side diff 1`] = `
"  {
-   "age": 33,
-   "objectData": {
-     "text": "text",
-   },
-   "nullData": null,
-   "arrayData": [
-     1,
-   ],
    "name": "some-name",
    "extraField": "some-data",
  }"
`;

exports[`diff-printer renders diff when left is missing properties or right has extra properties: 2) right side diff 1`] = `
"  {
-   "extraField": "some-data",
    "name": "some-name",
    "age": 33,
    "objectData": {
      "text": "text",
    },
    "nullData": null,
    "arrayData": [
      1,
    ],
  }"
`;

exports[`diff-printer renders diff when left is missing properties or right has extra properties: 3) left side diff exact 1`] = `
"  {
-   "age": 33,
-   "objectData": {
-     "text": "text",
-   },
-   "nullData": null,
-   "arrayData": [
-     1,
-   ],
    "name": "some-name",
+   "extraField": "some-data",
  }"
`;

exports[`diff-printer renders diff when left is missing properties or right has extra properties: 4) right side diff exact 1`] = `
"  {
-   "extraField": "some-data",
    "name": "some-name",
+   "age": 33,
+   "objectData": {
+     "text": "text",
+   },
+   "nullData": null,
+   "arrayData": [
+     1,
+   ],
  }"
`;

exports[`diff-printer renders diff when left is object with empty array and right is primitive: 1) left side diff 1`] = `
"- "text"
+ {
+   "emptyArray": [],
+ }"
`;

exports[`diff-printer renders diff when left is object with empty array and right is primitive: 2) right side diff 1`] = `
"- {
-   "emptyArray": [],
- }
+ "text""
`;

exports[`diff-printer renders diff when left is object with empty array and right is primitive: 3) left side diff exact 1`] = `
"- "text"
+ {
+   "emptyArray": [],
+ }"
`;

exports[`diff-printer renders diff when left is object with empty array and right is primitive: 4) right side diff exact 1`] = `
"- {
-   "emptyArray": [],
- }
+ "text""
`;

exports[`diff-printer renders diff when left is object with empty object and right is primitive: 1) left side diff 1`] = `
"- "text"
+ {
+   "emptyObject": {},
+ }"
`;

exports[`diff-printer renders diff when left is object with empty object and right is primitive: 2) right side diff 1`] = `
"- {
-   "emptyObject": {},
- }
+ "text""
`;

exports[`diff-printer renders diff when left is object with empty object and right is primitive: 3) left side diff exact 1`] = `
"- "text"
+ {
+   "emptyObject": {},
+ }"
`;

exports[`diff-printer renders diff when left is object with empty object and right is primitive: 4) right side diff exact 1`] = `
"- {
-   "emptyObject": {},
- }
+ "text""
`;

exports[`diff-printer renders diff when left is primitive and right is empty array: 1) left side diff 1`] = `
"- []
+ "text""
`;

exports[`diff-printer renders diff when left is primitive and right is empty array: 2) right side diff 1`] = `"  []"`;

exports[`diff-printer renders diff when left is primitive and right is empty array: 3) left side diff exact 1`] = `
"- []
+ "text""
`;

exports[`diff-printer renders diff when left is primitive and right is empty array: 4) right side diff exact 1`] = `"  []"`;

exports[`diff-printer renders diff when left is primitive and right is empty object: 1) left side diff 1`] = `
"- {}
+ "text""
`;

exports[`diff-printer renders diff when left is primitive and right is empty object: 2) right side diff 1`] = `
"- "text"
+ {}"
`;

exports[`diff-printer renders diff when left is primitive and right is empty object: 3) left side diff exact 1`] = `
"- {}
+ "text""
`;

exports[`diff-printer renders diff when left is primitive and right is empty object: 4) right side diff exact 1`] = `
"- "text"
+ {}"
`;

exports[`diff-printer renders diff when left is primitive and right is object: 1) left side diff 1`] = `
"- {
-   "data": "text",
- }
+ "text""
`;

exports[`diff-printer renders diff when left is primitive and right is object: 2) right side diff 1`] = `
"- "text"
+ {
+   "data": "text",
+ }"
`;

exports[`diff-printer renders diff when left is primitive and right is object: 3) left side diff exact 1`] = `
"- {
-   "data": "text",
- }
+ "text""
`;

exports[`diff-printer renders diff when left is primitive and right is object: 4) right side diff exact 1`] = `
"- "text"
+ {
+   "data": "text",
+ }"
`;

exports[`diff-printer renders diff when right array includes left array: 1) left side diff 1`] = `
"  [
    1,
    2,
-   3,
  ]"
`;

exports[`diff-printer renders diff when right array includes left array: 2) right side diff 1`] = `""`;

exports[`diff-printer renders diff when right array includes left array: 3) left side diff exact 1`] = `
"- [
-   3,
-   1,
-   2,
- ]
+ [
+   1,
+   2,
+ ]"
`;

exports[`diff-printer renders diff when right array includes left array: 4) right side diff exact 1`] = `
"- [
-   1,
-   2,
- ]
+ [
+   3,
+   1,
+   2,
+ ]"
`;

exports[`diff-printer renders diff with deep nested primitive value mismatch: 1) left side diff 1`] = `
"  {
    "lvl1": {
      "lvl2": {
        "lvl3": {
-         "mismatch": "date",
+         "mismatch": "data",
          "sameData": "txt",
        },
      },
    },
  }"
`;

exports[`diff-printer renders diff with deep nested primitive value mismatch: 2) right side diff 1`] = `
"  {
    "lvl1": {
      "lvl2": {
        "lvl3": {
-         "mismatch": "data",
+         "mismatch": "date",
          "sameData": "txt",
        },
      },
    },
  }"
`;

exports[`diff-printer renders diff with deep nested primitive value mismatch: 3) left side diff exact 1`] = `
"  {
    "lvl1": {
      "lvl2": {
        "lvl3": {
-         "mismatch": "date",
+         "mismatch": "data",
          "sameData": "txt",
        },
      },
    },
  }"
`;

exports[`diff-printer renders diff with deep nested primitive value mismatch: 4) right side diff exact 1`] = `
"  {
    "lvl1": {
      "lvl2": {
        "lvl3": {
-         "mismatch": "data",
+         "mismatch": "date",
          "sameData": "txt",
        },
      },
    },
  }"
`;

exports[`diff-printer renders diff with nested arrays with mismatches: 1) left side diff 1`] = `
"  {
    "data": [
      [
        1,
        2,
      ],
      [
        3,
        4,
      ],
-     [
-       3,
-       5,
-     ],
    ],
  }"
`;

exports[`diff-printer renders diff with nested arrays with mismatches: 2) right side diff 1`] = `
"  {
    "data": [
      [
        1,
        2,
      ],
      [
        3,
        5,
      ],
-     [
-       3,
-       4,
-     ],
    ],
  }"
`;

exports[`diff-printer renders diff with nested arrays with mismatches: 3) left side diff exact 1`] = `
"  {
    "data": [
      [
        1,
        2,
      ],
      [
        3,
-       5,
+       4,
      ],
    ],
  }"
`;

exports[`diff-printer renders diff with nested arrays with mismatches: 4) right side diff exact 1`] = `
"  {
    "data": [
      [
        1,
        2,
      ],
      [
        3,
-       4,
+       5,
      ],
    ],
  }"
`;

exports[`diff-printer renders diff with nested primitive and object mismatch: 1) left side diff 1`] = `
"  {
    "lvl1": {
      "lvl2": {
        "lvl3": {
-         "mismatch": {
-           "data": "txt",
-         },
+         "mismatch": "data",
          "sameData": "txt",
        },
      },
    },
  }"
`;

exports[`diff-printer renders diff with nested primitive and object mismatch: 2) right side diff 1`] = `
"  {
    "lvl1": {
      "lvl2": {
        "lvl3": {
-         "mismatch": "data",
+         "mismatch": {
+           "data": "txt",
+         },
          "sameData": "txt",
        },
      },
    },
  }"
`;

exports[`diff-printer renders diff with nested primitive and object mismatch: 3) left side diff exact 1`] = `
"  {
    "lvl1": {
      "lvl2": {
        "lvl3": {
-         "mismatch": {
-           "data": "txt",
-         },
+         "mismatch": "data",
          "sameData": "txt",
        },
      },
    },
  }"
`;

exports[`diff-printer renders diff with nested primitive and object mismatch: 4) right side diff exact 1`] = `
"  {
    "lvl1": {
      "lvl2": {
        "lvl3": {
-         "mismatch": "data",
+         "mismatch": {
+           "data": "txt",
+         },
          "sameData": "txt",
        },
      },
    },
  }"
`;

exports[`diff-printer renders diff with special characters in keys: 1) left side diff 1`] = `
"  {
-   "key-with-dash": "different",
+   "key-with-dash": "value1",
    "key.with.dots": "value2",
  }"
`;

exports[`diff-printer renders diff with special characters in keys: 2) right side diff 1`] = `
"  {
-   "key-with-dash": "value1",
+   "key-with-dash": "different",
    "key.with.dots": "value2",
  }"
`;

exports[`diff-printer renders diff with special characters in keys: 3) left side diff exact 1`] = `
"  {
-   "key-with-dash": "different",
+   "key-with-dash": "value1",
    "key.with.dots": "value2",
  }"
`;

exports[`diff-printer renders diff with special characters in keys: 4) right side diff exact 1`] = `
"  {
-   "key-with-dash": "value1",
+   "key-with-dash": "different",
    "key.with.dots": "value2",
  }"
`;

exports[`diff-printer should not render diff when left and right are equal: 1) left side diff 1`] = `""`;

exports[`diff-printer should not render diff when left and right are equal: 2) right side diff 1`] = `""`;

exports[`diff-printer should not render diff when left and right are equal: 3) left side diff exact 1`] = `""`;

exports[`diff-printer should not render diff when left and right are equal: 4) right side diff exact 1`] = `""`;
